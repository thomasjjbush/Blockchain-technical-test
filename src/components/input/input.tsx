import React, { ChangeEvent, FC, ReactElement } from 'react';
import styled from 'styled-components';
import { SearchInputProps as Props } from '../../../types';

const Container = styled.span`
    border: solid 2px ${(props): string => props.theme.colors.offset};
    border-radius: 5px;
    overflow: hidden;
`;

const Button = styled.button`
    border: none;
    height: 100%;
    padding: 0 10px;
`;

const Input = styled.input`
    border: none;
    height: 100%;
    padding: 0 10px;
`;

export const SearchInput: FC<Props> = ({ onChange, value, placeholder }: Props): ReactElement => {
    return (
        <Container>
            <Input
                data-test-id="search"
                onChange={(e: ChangeEvent<HTMLInputElement>) => onChange(e.currentTarget.value)}
                placeholder={placeholder}
                value={value}
                type="text"
            />
            <Button data-test-id="clear" disabled={!value} onClick={() => onChange('')}>
                x
            </Button>
        </Container>
    );
};
