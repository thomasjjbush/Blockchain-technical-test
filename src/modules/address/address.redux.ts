import { AnyAction } from 'redux';
import { Action, Address, AddressActions } from '../../../types';
import { fetchData } from '../../utils/fetch/fetch-data';

export const loadAddress = (address: string, multiplier: number): Action => async (dispatch, getState) => {
    const offset = multiplier * getState().address.tx_per_page;
    try {
        const res = await fetchData<Address>(`rawaddr/${address}`, { offset, limit: 10 });
        return dispatch({ type: AddressActions.LOADED, payload: res });
    } catch (err) {
        // REMOVE THIS NESTED TRY CATCH
        try {
            const multiRes = await fetchData<any>('multiaddr', { active: address, offset, limit: 10 });
            return dispatch({ type: AddressActions.LOADED, payload: { ...multiRes.addresses[0], txs: multiRes.txs } });
        } catch (err) {
            return dispatch({ type: AddressActions.ERROR });
        }
    }
};

export const addressReducer = (state: Address = { tx_per_page: 10 }, action: AnyAction): Address => {
    switch (action.type) {
        case AddressActions.ERROR:
            return { ...state, error: true };
        case AddressActions.LOADED:
            return { ...state, ...(action.payload as Address) };
        default:
            return state;
    }
};
