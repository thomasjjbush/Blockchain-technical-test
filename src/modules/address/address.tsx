import React, { FC, ReactElement, useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useLocation, useParams } from 'react-router-dom';
import styled from 'styled-components';
import { loadAddress } from './address.redux';
import { MultiPageButtons } from '../../components/multi-page-buttons/multi-page-buttons';
import { SearchInput } from '../../components/input/input';
import { Address, Store, StyledProps } from '../../../types';
import { Transaction } from '../../components/transaction/transaction';

const Block = styled.div`
    background-color: white;
    margin-bottom: 30px;
    padding: 30px;

    &:last-of-type {
        margin-bottom: 0;
    }
`;

const Container = styled.div`
    max-width: 1024px;
    margin: auto;
    padding: 30px;
`;

const Title = styled.h1<StyledProps>`
    color: ${(props): string => props.theme.colors.brand};
    margin: 0;
`;

const Address: FC = (): ReactElement => {
    const dispatch = useDispatch();
    const address = useSelector<Store, Address>((state) => state.address);
    const page = Math.max(Number(new URLSearchParams(useLocation().search).get('page')), 1);
    const { slug } = useParams<{ slug: string }>();
    const [search, setSearch] = useState('');

    const changePage = useCallback((i: number): Partial<Location> => ({ search: `?page=${i}` }), []);

    useEffect(() => {
        setSearch('');
        dispatch(loadAddress(slug, page - 1));
    }, [page, slug]);

    if (address.error) return <p>error</p>;
    if (!address.address) return <p>loading</p>;

    const numberOfPages = Math.floor(address.n_tx / address.tx_per_page);

    return (
        <Container>
            <Block>
                <Title>Address</Title>
                <p>Transactions: {address.n_tx}</p>
                <p>Total received: {address.total_received}</p>
                <p>Total sent: {address.total_sent}</p>
                <p>Final Balance: {address.final_balance}</p>
            </Block>
            <Block>
                <Title>
                    Transactions (Page {page}/{numberOfPages})
                </Title>
                <SearchInput onChange={setSearch} placeholder="Filter by hash" value={search} />
                {address.txs
                    .filter(({ hash }) => hash.includes(search))
                    .map(({ fee, hash }) => (
                        <Transaction key={hash} fee={fee} hash={hash} search={search} />
                    ))}
                <MultiPageButtons
                    currentPage={page}
                    numberOfButtons={5}
                    numberOfPages={numberOfPages}
                    skip={10}
                    to={changePage}
                />
            </Block>
        </Container>
    );
};

export default Address;
