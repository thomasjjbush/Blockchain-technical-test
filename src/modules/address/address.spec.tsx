import React from 'react';
import { useSelector } from 'react-redux';
import { mount, shallow } from 'enzyme';
import Address from './address';
import { loadAddress } from './address.redux';
import { MultiPageButtons } from '../../components/multi-page-buttons/multi-page-buttons';
import { SearchInput } from '../../components/input/input';

jest.mock('react-router-dom', () => ({
    useLocation: jest.fn().mockReturnValue({ search: 'page=2' }),
    useParams: jest.fn().mockReturnValue({ slug: 'slug' }),
}));

jest.mock('react-redux', () => ({
    useDispatch: jest.fn().mockReturnValue(jest.fn()),
    useSelector: jest.fn().mockReturnValue({ tx_per_page: 10 }),
}));

jest.mock('./address.redux', () => ({
    loadAddress: jest.fn().mockReturnValue({ type: 'MOCK_ACTION' }),
}));

jest.mock('../../components/multi-page-buttons/multi-page-buttons', () => ({
    MultiPageButtons: () => 'MultiPageButtons',
}));

describe('Address', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should return loading text if no address data is available', () => {
        const wrapper = shallow(<Address />);
        expect(wrapper.find('p').text()).toBe('loading');
    });

    it('should render error text if there is an error', () => {
        (useSelector as jest.Mock).mockReturnValue({ error: true });
        const wrapper = shallow(<Address />);
        expect(wrapper.find('p').text()).toBe('error');
    });

    it('should invoke loadAddress action on mount', () => {
        mount(<Address />);
        expect(loadAddress).toHaveBeenCalledTimes(1);
        expect(loadAddress).toHaveBeenCalledWith('slug', 1);
    });

    describe('with data', () => {
        beforeEach(() => {
            (useSelector as jest.Mock).mockReturnValue({
                address: 'lordSouron',
                final_balance: 29,
                n_tx: 20,
                total_received: 55,
                total_sent: 33,
                tx_per_page: 10,
                txs: [
                    {
                        balance: 20,
                        fee: 10,
                        hash: '24423423',
                        result: 14916,
                        size: 300,
                        time: 112313,
                    },
                    {
                        balance: 20,
                        fee: 10,
                        hash: '24423423SEARCHABLE',
                        result: 14916,
                        size: 300,
                        time: 112313,
                    },
                ],
            });
        });

        it('should match snapshot', () => {
            expect(shallow(<Address />)).toMatchSnapshot();
        });

        it('should filter based on hash', () => {
            const wrapper = shallow(<Address />);
            expect(wrapper.find({ 'data-test-id': 'transaction' }).length).toBe(2);

            wrapper.find(SearchInput).prop('onChange')('SEARCHABLE');
            expect(wrapper.find({ 'data-test-id': 'transaction' }).length).toBe(1);
            expect(wrapper.find({ 'data-test-id': 'hash' }).find('span').text()).toBe('SEARCHABLE');
        });

        it('should pass the correct cb to multi page buttons component', () => {
            const wrapper = shallow(<Address />);
            expect(wrapper.find(MultiPageButtons).prop('to')(5)).toEqual({ search: '?page=5' });
        });
    });
});
