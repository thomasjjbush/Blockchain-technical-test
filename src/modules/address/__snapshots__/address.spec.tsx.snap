// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address with data should match snapshot 1`] = `
<styled.div
  styles="
    max-width: 1024px;
    margin: auto;
    padding: 30px;
"
>
  <styled.div
    styles="
    background-color: white;
    margin-bottom: 30px;
    padding: 30px;

    &:last-of-type {
        margin-bottom: 0;
    }
"
  >
    <styled.h1
      styles="
    color: theme.colors.brand;
    margin: 0;
"
    >
      Address
    </styled.h1>
    <p>
      Transactions: 
      20
    </p>
    <p>
      Total received: 
      55
    </p>
    <p>
      Total sent: 
      33
    </p>
    <p>
      Final Balance: 
      29
    </p>
  </styled.div>
  <styled.div
    styles="
    background-color: white;
    margin-bottom: 30px;
    padding: 30px;

    &:last-of-type {
        margin-bottom: 0;
    }
"
  >
    <styled.h1
      styles="
    color: theme.colors.brand;
    margin: 0;
"
    >
      Transactions (Page 
      2
      /
      2
      )
    </styled.h1>
    <SearchInput
      onChange={[Function]}
      placeholder="Filter by hash"
      value=""
    />
    <Transaction
      fee={10}
      hash="24423423"
      key="24423423"
      search=""
    />
    <Transaction
      fee={10}
      hash="24423423SEARCHABLE"
      key="24423423SEARCHABLE"
      search=""
    />
    <MultiPageButtons
      currentPage={2}
      numberOfButtons={5}
      numberOfPages={2}
      skip={10}
      to={[Function]}
    />
  </styled.div>
</styled.div>
`;
